### PR TITLE
MNT Add travis file

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripeltd/product-issues/issues/437

PHP 7.4 + MySQL 7.4

### CODE REVIEW

**Comment by emteknetnz** [comment]
Reference build with php 7.3
https://travis-ci.com/github/silverstripe/cwp-recipe-basic/builds/227197034

Recent build with php 7.4 and deprecated errors surpressed
https://travis-ci.com/github/silverstripe/cwp-recipe-basic/builds/227687722

Differences
[Job 1](https://travis-ci.com/github/silverstripe/cwp-recipe-basic/jobs/510267889)
- same failures as php7.3, until it stopped midway through with error `Trying to access array offset on value of type bool`

[Job 2](https://travis-ci.com/github/silverstripe/cwp-recipe-basic/jobs/508791683)
- same failures as php7,3

[Job 3](https://travis-ci.com/github/silverstripe/cwp-recipe-basic/jobs/510267891)
- same failures as php7.3

[Job 4](https://travis-ci.com/github/silverstripe/cwp-recipe-basic/jobs/510267892)
- same failures as php7.3 - green test though has user error `ERROR [User Error]: Uncaught Exception: just excepted`

[Job 5](https://travis-ci.com/github/silverstripe/cwp-recipe-basic/jobs/510267893)
- same failures as php7.3

**Comment by brynwhyman** [comment]
Pretty keen to see this merged so we can in the future have a quick way to see if compatibility with these requirements cause issues down the line

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  # Numeric branch - 4 / 4.8 => 4.x-dev / 4.8.x-dev
+  # Tag - 1.2.3 / 1.2.3-beta1 => 1.2.3 / 1.2.3-beta1
+  # Non-numeric branch - master / some-branch => dev-master / dev-some-branch
+  # $TRAVIS_BRANCH documentation - https://docs.travis-ci.com/user/environment-variables/#default-environment-variables
+  - if [[ "$TRAVIS_BRANCH" =~ ^[1-9]$ ]] || [[ "$TRAVIS_BRANCH" =~ ^[1-9]\.[0-9]+$ ]]; then export COMPOSER_ROOT_VERSION="${TRAVIS_BRANCH}.x-dev"; elif [[ "$TRAVIS_BRANCH" =~ ^[1-9]\.[0-9]+\.[0-9]+ ]]; then export COMPOSER_ROOT_VERSION="${TRAVIS_BRANCH}"; else export COMPOSER_ROOT_VERSION="dev-${TRAVIS_BRANCH}"; fi
+
+  # PHPENV
+  - phpenv rehash
+  - phpenv config-rm xdebug.ini || true
+  - if [[ $PHPUNIT_COVERAGE_TEST ]]; then memlimit='8192M'; else memlimit='4096M'; fi
```
Comment (maxime-rainville) [inline .travis.yml]: Why is this needed ... we don't have any matrix entry with `PHPUNIT_COVERAGE_TEST`

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  - echo "memory_limit = ${memlimit}" >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
+  - echo 'always_populate_raw_post_data = -1' >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
+
+  # COMPOSER
+  - if [ $COMPOSER_VERSION ]; then composer self-update "$COMPOSER_VERSION"; elif [ $(php -r 'echo (int) version_compare(phpversion(), "7.3.0", "<=");') = "1" ]; then composer self-update 1.9.3; else composer self-update; fi
```
Comment (maxime-rainville) [inline .travis.yml:64]: All the matrix entries are running PHP 7.4

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  # silverstripe-installer - includes files required for minimal site setup e.g. Page, PageController
+  - composer create-project silverstripe/installer:^3 ~/myinstaller
+  - cp -R ~/myinstaller/assets .
+  - cp -R ~/myinstaller/mysite .
+  - cp -R ~/myinstaller/themes .
+  - cp ~/myinstaller/.htaccess .
+  - cp ~/myinstaller/index.php .
+
+  # turn off php 7.4 deprecation warnings
+  - echo "error_reporting(E_ALL ^ E_DEPRECATED);" >> mysite/_config.php
```
Comment (maxime-rainville) [inline .travis.yml]: I think it would make more sense to add this to the `php.ini` file like this: https://docs.travis-ci.com/user/languages/php/#custom-php-configuration

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  - echo "define('SS_TRUSTED_PROXY_IPS', '*');" >> _ss_environment.php
+  - echo "define('SS_DATABASE_SERVER', '127.0.0.1');" >> _ss_environment.php
+  - echo "define('SS_DATABASE_USERNAME', 'root');" >> _ss_environment.php
+  - echo "define('SS_DATABASE_PASSWORD', '');" >> _ss_environment.php
+  - echo "define('SS_DATABASE_CHOOSE_NAME', 1);" >> _ss_environment.php
+  - echo "define('SS_DEFAULT_ADMIN_USERNAME', 'admin');" >> _ss_environment.php
+  - echo "define('SS_DEFAULT_ADMIN_PASSWORD', 'password');" >> _ss_environment.php
+  - echo "define('SS_HOST', 'localhost');" >> _ss_environment.php
+  - echo "global \$_FILE_TO_URL_MAPPING;" >> _ss_environment.php
+  - echo "\$_FILE_TO_URL_MAPPING['$PWD'] = 'http://localhost:8000';" >> _ss_environment.php
```
Comment (maxime-rainville) [inline .travis.yml:93]: Is this needed? I ran travis build on CMS3 before and didn't need all this stuff.

**Review by maxime-rainville — CHANGES_REQUESTED** [review]
This seems to be more convoluted than it needs to be.

We use to have well define way of setting up builds for CMS 3 modules.

Is this repo not working any more? https://github.com/silverstripe/silverstripe-travis-support

You can look at these travis config for examples:
- https://github.com/silverstripe/silverstripe-cms/blob/3/.travis.yml
- https://github.com/webtorque/silverstripe-notifications/blob/master/.travis.yml

Build is currently failing ... is that expected at this stage.

File: .travis.yml (hunk 4 above)
Comment (emteknetnz) [reply .travis.yml:93]: Yes needed.

Chances are it worked because the old builds were using this rather outdated helper https://github.com/silverstripe/silverstripe-travis-support which is very out of date and we shouldn't update it just for the sake of a single build, since it's been entirely superseded by https://github.com/silverstripe/silverstripe-travis-shared

File: .travis.yml (hunk 2 above)
Comment (emteknetnz) [reply .travis.yml:64]: I had it running php 7.3 previously - that's how I validated that PHP 7.3 and PHP 7.4 behave the same once deprecation warnings are removed

I think it's worth keeping this as composer 1 is still around.  It also matches what's in travis-shared

File: .travis.yml (hunk 1 above)
Comment (emteknetnz) [reply .travis.yml]: Removed the conditional

File: .travis.yml (hunk 3 above)
Comment (emteknetnz) [reply .travis.yml]: Updated

**Comment by emteknetnz** [comment]
> Build is currently failing ... is that expected at this stage.

Yup there's known issues which aren't being addressed in this PR.  This is just to get some level of CI going

**Review by maxime-rainville — APPROVED** [review]
My nitpicking has been addressed. I've confirmed the build run has expected on travis.
